### PR TITLE
RFC: apigw-lambda library WIP

### PR DESCRIPTION
**Pull request by jvarho**
This is still very much work in progress, but I'm pushing this in case someone has suggestions. This only includes handler building and default/api page handling code at the moment, plus a bunch of tests ported from the lambda-at-edge library. And even in the default handler ISR support is still missing. (And deploying it by hand still requires some manual fixing.)

I'm currently proceeding under the following assumptions:

* Next.js 10.2.3+ support only – no serverless trace target (seems to use more space nowadays)
* Currently only one lambda, handles both API pages and others, since they will share the chunks dir anyway.
* Regeneration events will be handled in the same lambda, just as different events. (Again, same code.)
* Image handler will get its own lambda when I get to it, since it shares almost no code with the rest.

What I'm less sure about is the higher level implementation.

Should I add support in the serverless component? How? That seems like a bit of a dead end, considering it's using beta components and the GA components are a completely different beast. But I'm happy to do that if it's seen as best. (I'm not sure I personally want to use the SAAS model GA components are based on.)

Should I add support in the CDK construct? Or create a separate CDK package? I need to study up a bit on CDK first if this is the way to go.

Ps. this is only about 1k lines of implementing code. Most of the diff is tests and yarn lock.

TODO:

- [x] support api pages in default handler
- [x] reduce test code duplication
- [ ] reduce implementation code duplication
- [ ] support regeneration in default handler
- [ ] create image handler
- [ ] CDK v2 support for this
- [ ] documentation

### CODE REVIEW

**Comment by dphang** [comment]
Will look at this during the next week. A few things:

* I think we can combine API and default handler if Next.js code splitting is more optimized now (it will also help with issues like doing cross-rewrites between default and API routes)
* Yes, serverless GA upgrade has been long overdue though with things like CDK etc not sure if we need to keep going that direction (also yes they ofc will want to do SaaS model), prefer to keep it as open as possible and also avoid us having to manually code against the AWS SDK. I think you could start with CDK (v2)?

**Comment by dphang** [comment]
Had another look. While it's great, I do see there is still some duplicated code in the handler and also the tests are just ported and duplicated/adapted from lambda-at-edge?

I guess the vision I had was that the core package should have a base generic handler and as mentioned before, all existing default/api handler tests go there (testing the core logic independent of S3/CloudFront like it is now), and then lambda-at-edge / apigw-lambda / other implementations would just essentially wrap the generic handlers and add relevant logic (i.e CloudFront events, API GW, etc.).

Otherwise it will be hard to maintain two sets of implementations that share logic/tests. So I guess it can be fine for an initial proof of concept but probably we shouldn't continue on this path, but plan to refactor and simplify it soon before it becomes ready for a stable release.

I'll take a look at the other code itself soon.

**Comment by jvarho** [comment]
> Had another look. While it's great, I do see there is still some duplicated code in the handler and also the tests are just ported and duplicated/adapted from lambda-at-edge?

Thanks, yes I agree there is still some duplication that could be avoided. However, I'm not yet sure what to about it.

Some of the shared logic I can still move to libs/core. But some is very AWS specific, so if the idea is to eventually support other environments, that might not be the best place for it. But once this is more feature complete I'll deduplicate somehow for sure.

> I guess the vision I had was that the core package should have a base generic handler and as mentioned before, all existing default/api handler tests go there[...]

The issue with sharing tests is that the lambda-at-edge tests are in large part very implementation specific. They (have to) test that the handlers return either responses or S3 requests, as appropriate. Whereas here there is no response handler, so the tests are that the correct S3 request is made.

But yeah, I should probably reduce the kinds of things that are tested at this level and move all the more specific testing to happen in libs/core. I.e. if this level doesn't care whether it's a dynamic HTML page or a static SSG page that's requested from S3, then there's no reason to test both.

(Personally, I find the large number of tests to help with maintenance more than it hurts. But the implementation should certainly duplicate as little code as possible.)

**Comment by jvarho** [comment]
As an aside:

Yes, moving lambda-at-edge to use a single handler + S3 client would allow more code sharing and sharing tests at a more abstract level. But while working on this I've realized that means a 1MB hard limit on public/ files. That is ok for some applications, but I'm not sure it's a good idea to break proper file support. At least with apigw + lambda the limit is 6MB.

**Comment by dphang** [comment]
> As an aside:
> 
> 
> 
> Yes, moving lambda-at-edge to use a single handler + S3 client would allow more code sharing and sharing tests at a more abstract level. But while working on this I've realized that means a 1MB hard limit on public/ files. That is ok for some applications, but I'm not sure it's a good idea to break proper file support. At least with apigw + lambda the limit is 6MB.

I wasn't planning to use s3 client for all files, for public files / static pages I am thinking the generic core handler would just return a generic response instructing to get a static file. Then lambda@edge would be able to return a response to tell CloudFront to serve the file (no 1 MB limit)

I think s3 client is mostly needed to replace use of the response handler, where we need to modify the s3 response or status code of static pages. I think most people would have small page sizes?

**Comment by jvarho** [comment]
> > As an aside:
> > Yes, moving lambda-at-edge to use a single handler + S3 client would allow more code sharing and sharing tests at a more abstract level. But while working on this I've realized that means a 1MB hard limit on public/ files. That is ok for some applications, but I'm not sure it's a good idea to break proper file support. At least with apigw + lambda the limit is 6MB.
> 
> I wasn't planning to use s3 client for all files, for public files / static pages I am thinking the generic core handler would just return a generic response instructing to get a static file. Then lambda@edge would be able to return a response to tell CloudFront to serve the file (no 1 MB limit)

Makes sense, probably the same with prerendered pages? No such option here however. So tests have to differ quite a lot.

**Comment by dphang** [comment]
> > > As an aside:
> > > Yes, moving lambda-at-edge to use a single handler + S3 client would allow more code sharing and sharing tests at a more abstract level. But while working on this I've realized that means a 1MB hard limit on public/ files. That is ok for some applications, but I'm not sure it's a good idea to break proper file support. At least with apigw + lambda the limit is 6MB.
> > 
> > 
> > I wasn't planning to use s3 client for all files, for public files / static pages I am thinking the generic core handler would just return a generic response instructing to get a static file. Then lambda@edge would be able to return a response to tell CloudFront to serve the file (no 1 MB limit)
> 
> Makes sense, probably the same with prerendered pages? No such option here however. So tests have to differ quite a lot.

Yes, I see you need to use the s3 client here always, basically I was trying to see if you can share the core logic of [rendering SSR pages + static files]. For static files, the core logic just says "Get a static file located at this path" but doesn't know how to get it. It's up to the platform-specific handler to know how to retrieve it.

* For lambda-at-edge: it can return a CloudFrontRequest (to tell CloudFront to retrieve from the S3 origin) or use s3-client (directly gets S3 within lambda, could potentially avoid overhead of CloudFront requesting S3, at a tradeoff of 1 MB response limit and additional Lambda time cost).
* For lambda + apigw: it would always need to use s3-client, though thankfully there is a 6 MB response limit, so it's fine.

I think you already have a good start with the `staticRequest` in lambda-at-edge, when moving to core, that just needs to be detached from CloudFront concepts

**Comment by jvarho** [comment]
> I think you already have a good start with the `staticRequest` in lambda-at-edge, when moving to core, that just needs to be detached from CloudFront concepts

You mean passing a callback like that to core code as a parameter? I guess that would work, but it needs to be able to handle both the case where the request may fail (dynamic SSG with fallback) and where it won't. And assuming that e.g. public files will never fail means no nice 404 page if they do (e.g. during deployment).

If that logic is moved to core, there also needs to be a way to store pages, so that's another callback I guess.

I'll have a look how complex this would be and whether I can get it done before finishing up this or leave it as a future improvement.

File: packages/libs/apigw-lambda/src/default-handler.ts [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+    if (s3Response.ContentType) {
+      res.setHeader("Content-Type", s3Response.ContentType);
+    }
+    if (s3Response.CacheControl) {
+      res.setHeader("Cache-Control", s3Response.CacheControl);
+    }
+    res.write(bodyString);
+    return true;
+  } catch (error) {
+    return false;
```
[the remark this answers is not in the document]
Comment (jvarho) [reply packages/libs/apigw-lambda/src/default-handler.ts:123]: It's an expected exception if the file does not exist and common in fallback. I don't think it makes sense to warn about it?

I'll see if I can warn on other exceptions.

File: packages/libs/apigw-lambda/src/types.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    enableHTTPCompression?: boolean;
+    logLambdaExecutionTimes?: boolean;
+    region: string;
+  };
+
+export type RequestEvent = APIGatewayProxyEventV2;
+
+export type EventResponse = APIGatewayProxyStructuredResultV2;
+
+export type PreRenderedManifest = {
```
[the remark this answers is not in the document]
Comment (jvarho) [reply packages/libs/apigw-lambda/src/types.ts:20]: Some yes, others differ between apigw/edge at least for now.

**Comment by jvarho** [comment]
Made an initial rebase on top of current master. Still work to do to even integrate all the changes in the past couple of months.

Debugging some local compile hangs at the moment...

**Comment by dphang** [comment]
@jvarho FYI I'm also refactoring the core handler logic to make it platform agnostic in here: https://github.com/serverless-nextjs/serverless-next.js/pull/1747 so it'll be easier to support Lambda, Azure, Cloudflare, etc. in the future. 

Basically I created generic default/regeneration handlers so far which require a generic interface called `PlatformClient` which is supposed to interact with the backing stores/queues. Right now there is just an `AwsPlatformClient` which basically moved all the S3/SQS logic into the `@sls-next/aws-common` library. And Default + Regeneration Lambdas in Lambda@Edge now have V2 and basically just create a thin wrapper around the platform-agnostic handlers.

No tests yet since it's experimental at this point, but will work on adding tests soon (and also work on refactoring imageOptimizer out into the core package).

Hopefully it might be reusable for this?

**Comment by dphang** [comment]
@jvarho thanks for all your work on this! I basically adapted parts of this + built on top of the generic handlers and moved the logic into https://github.com/serverless-nextjs/serverless-next.js/tree/master/packages/libs/lambda

Right now I am writing deployment logic (using CDK for Terraform / CDK) and testing it on my AWS account, I hope to launch it in developer preview soon.

So I think we can close this in favor of the existing work that's already been done. For unit tests, we don't want to duplicate tests, so separate will work to move Lambda@Edge handler tests into the core package and I'll add lightweight unit tests to just test the Lambda/APIGWv2 compatibility layer (and of course e2e tests later).

**Comment by jvarho** [comment]
> @jvarho thanks for all your work on this! I basically adapted parts of this + built on top of the generic handlers and moved the logic into https://github.com/serverless-nextjs/serverless-next.js/tree/master/packages/libs/lambda
>
> Right now I am writing deployment logic (using CDK for Terraform for now) and testing it on my AWS account, I hope to launch it in developer preview soon.

Sounds good! Sorry for not following up on this PR earlier, life has got in the way.
